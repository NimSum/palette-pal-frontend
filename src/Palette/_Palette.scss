
.Palette {
  display: flex;
  flex-direction: column;
  margin: 15px 20px;   
  width: 16%;
  text-decoration: none;
  position: relative;

  .fa-times-circle {
    display: none;
  }

  &:hover {
    .fa-times-circle {
      display: block;
      position: absolute;
      top: -5px;
      right: -15px;
      font-size: 20px;
      color: $btn-color;
      background: white;
      border-radius: 100%;
      
      &:hover {
        color: darkred;
        cursor: pointer;
      }
    }
  }

  .color {
    width: 100%;
    min-width: 150px;
    height: 40px;
    display: flex;
    align-items: center;
    justify-content: center;

    .color-hex {
      font-size: 14px;
      color: white;
      mix-blend-mode: difference;
      opacity: .7;
      // filter: contrast(10%);
    }
  }

  .palette-title {
    font-style: italic;
    font-size: 13px;
    color: rgb(96, 96, 96);
    margin: 8px 0 10px 0;
  }
}
