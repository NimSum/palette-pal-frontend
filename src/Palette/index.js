import React from 'react';
import colorFormatter from '../utils/colorFormatter';
import colorContraster from '../utils/colorContraster';

function Palette(props) {
  const colorDivs = [];

  // const formatted = colorFormatter(props.data, props.format)
  
  for (let i = 1; i <= 5; i++) {
    const color = props.data[`color_${i}`]
    colorDivs.push(
      <div className="color" key={props.data.name + color} style={{ backgroundColor: color }}>
        <p className="color-hex" style={{ color: colorContraster(color) }}>{colorFormatter(color, props.format)}</p>
      </div>
    )
  }

  return (
    <div className="Palette" >
      <i className="fas fa-times-circle" onClick={() => {
        props.updatePaletteData({ ...props.data, project_id: props.projectID }, 'delete');
      }}/>
      {colorDivs}
      <p
        className="palette-title"
        contentEditable
        suppressContentEditableWarning
        onBlur={e => props.updatePaletteData({
          palette_name: e.target.textContent,
          id: props.data.id,
          project_id: props.projectID
        }, 'update')}
        onKeyDown={e => {
          if (e.keyCode === 13 || (e.target.textContent.length >= 15 && e.keyCode !== 8)) {
            e.preventDefault()
            props.updatePaletteData({
              palette_name: e.target.textContent,
              id: props.data.id,
              project_id: props.projectID
            }, 'update')
          }
        }}
      >
        {props.data.name}
      </p>
  </div>
  );
};

export default Palette;