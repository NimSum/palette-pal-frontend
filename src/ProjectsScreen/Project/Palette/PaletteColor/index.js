import React, { Component } from 'react';
import { PhotoshopPicker } from 'react-color';
import colorFormatter from '../../../../utils/colorFormatter';
import colorContraster from '../../../../utils/colorContraster';

class PaletteColor extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showPicker: false,
      update: ''
    }
  }

  handleChangeComplete = color => {
    this.setState({ update: color.hex })
  }

  updateColor = () => {
    const { palette_id, palette_name, project_id } = this.props;
    this.props.updatePaletteData({id: palette_id, [this.props.id]: this.state.update, project_id, palette_name}, 'update');
    this.setState({ update: '', showPicker: false })
  }

  render() {
    const picker = this.state.showPicker ? (
      <PhotoshopPicker
        header="Choose new color"
        className="PhotoshopPicker"
        color={this.state.update || this.props.color}
        onChangeComplete={this.handleChangeComplete}
        onAccept={this.updateColor}
        onCancel={() => this.setState({ showPicker: !this.state.showPicker, update: '' })}
      />
    ) : null;

    return (
      <div
        className="PaletteColor"
        key={this.props.id}
        style={{ backgroundColor: this.props.color }}
      >
        {picker}
        <p
          className="color-hex"
          style={{ color: colorContraster(this.props.color) }}
        >
          {colorFormatter(this.props.color, this.props.format)}
          <i
            className="fas fa-pencil-alt"
            onClick={() => this.setState({showPicker: !this.state.showPicker})}
          />
        </p>
      </div>
    );
  }
}

export default PaletteColor;