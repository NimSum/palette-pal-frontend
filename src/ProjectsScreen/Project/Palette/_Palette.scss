
.Palette {
  display: flex;
  flex-direction: column;
  margin: 15px 20px;   
  width: 16%;
  text-decoration: none;
  position: relative;

  .fa-times-circle {
    display: none;
  }

  &:hover {
    .fa-times-circle {
      display: block;
      position: absolute;
      top: -5px;
      right: -15px;
      font-size: 20px;
      color: $btn-color;
      background: white;
      border-radius: 100%;
      opacity: .85;
      
      &:hover {
        color: darkred;
        cursor: pointer;
      }
    }
  }

  .palette-title {
    font-style: italic;
    font-size: 13px;
    color: rgb(96, 96, 96);
    margin: 8px 0 10px 0;
  }
}
