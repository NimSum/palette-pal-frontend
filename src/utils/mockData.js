const mockData = {
  mockDetailedProjects: [
    {
      "project_name": "Project ONE",
      "project_id": 1,
      "palette_name": "Palette ONE",
      "palette_id": 1,
      "color_1": "#000000",
      "color_2": "#000000",
      "color_3": "#319547",
      "color_4": "#000000",
      "color_5": "#834267"
    },
    {
      "project_name": "Project TWO",
      "project_id": 2,
      "palette_name": "Palette TWO",
      "palette_id": 2,
      "color_1": "#010101",
      "color_2": "#925578",
      "color_3": "#319547",
      "color_4": "#796876",
      "color_5": "#834267"
    },
    {
      "project_name": "Project ONE",
      "project_id": 2,
      "palette_name": "Palette THREE",
      "palette_id": 3,
      "color_1": "#123456",
      "color_2": "#654321",
      "color_3": "#245216",
      "color_4": "#784895",
      "color_5": "#235686"
    }
  ],
  mockProjects: [
    {
      "id": 1,
      "project_name": "Project ONE",
      "created_at": "2019-07-05T21:39:51.436Z",
      "updated_at": "2019-07-05T21:39:51.436Z"
    },
    {
      "id": 2,
      "project_name": "Project TWO",
      "created_at": "2019-07-05T21:40:51.436Z",
      "updated_at": "2019-07-05T21:40:51.436Z"
    }
  ],
  mockPalettes: [
    {
      "id": 1,
      "palette_name": "Test ONE",
      "project_id": 1,
      "color_1": "#010101",
      "color_2": "#925578",
      "color_3": "#319547",
      "color_4": "#796876",
      "color_5": "#834267",
      "created_at": "2019-07-05T21:39:57.227Z",
      "updated_at": "2019-07-05T21:39:57.227Z"
    },
    {
        "id": 2,
        "palette_name": "Test TWO",
        "project_id": 2,
        "color_1": "#000000",
        "color_2": "#000000",
        "color_3": "#319547",
        "color_4": "#000000",
        "color_5": "#834267",
        "created_at": "2019-07-06T01:24:45.881Z",
        "updated_at": "2019-07-06T01:24:45.881Z"
    }
  ]
}

export default mockData;