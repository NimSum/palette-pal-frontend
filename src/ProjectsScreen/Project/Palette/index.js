import React from 'react';
import PaletteColor from './PaletteColor';

function Palette(props) {
	const paletteColors = [];

  for (let i = 1; i <= 5; i++) {
    const color = props.data[`color_${i}`];
    
    paletteColors.push(<PaletteColor
      color={color}
      palette_id={props.data.id}
      id={`color_${i}`}
      key={`color_${i}`}
      project_id={props.projectID}
      updatePaletteData={props.updatePaletteData}
      format={props.format}
    />);
	}

	return (
		<div className="Palette">
			<i
				className="fas fa-times-circle"
				onClick={() => {
					props.updatePaletteData({ ...props.data, project_id: props.projectID }, 'delete');
				}}
			/>
			{paletteColors}
			<p
				className="palette-title"
				contentEditable
				suppressContentEditableWarning
				onBlur={e =>
					props.updatePaletteData(
						{
							palette_name: e.target.textContent,
							id: props.data.id,
							project_id: props.projectID
						},
						'update'
					)}
				onKeyDown={e => {
					if (e.keyCode === 13 || (e.target.textContent.length >= 15 && e.keyCode !== 8)) {
						e.preventDefault();
						props.updatePaletteData(
							{
								palette_name: e.target.textContent,
								id: props.data.id,
								project_id: props.projectID
							},
							'update'
						);
					}
				}}>
				{props.data.name}
			</p>
		</div>
	);
}

export default Palette;
