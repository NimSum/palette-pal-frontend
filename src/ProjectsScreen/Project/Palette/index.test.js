import React from 'react';
import Palette from './index';
import { shallow } from 'enzyme';
import mockData from '../../../utils/mockData';

describe('Palette', () => {
  let wrapper;
  let instance;
  let mockPalette = mockData.mockPalettes[0];
  const mockUpdatePaletteData = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<Palette
      data={mockPalette}
      format={'hex'}
      projectID={12}
      updatePaletteData={mockUpdatePaletteData}
    />);
    instance = wrapper.instance();
  })
  
  it('should match component Palette snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call updatePaletteData when the delete icon is clicked', () => {
    wrapper.find('.fa-times-circle').simulate('click');
    
    expect(mockUpdatePaletteData).toHaveBeenCalled();
  })

  it('should call updatePaletteData when a content editable field focus is left', () => {
    const mockEvent = { target: { }, preventDefault: jest.fn(), keyCode: 13 }
    wrapper.find('.palette-title').simulate('blur', mockEvent);

    expect(mockUpdatePaletteData).toHaveBeenCalled();
  })

  it('should call updatePaletteData when a content editable field is in focus and the delete key is pressed', () => {
    const mockEvent = { target: { }, preventDefault: jest.fn(), keyCode: 13 }

    wrapper.find('.palette-title').simulate('keydown', mockEvent);

    expect(mockUpdatePaletteData).toHaveBeenCalled();
  })
})