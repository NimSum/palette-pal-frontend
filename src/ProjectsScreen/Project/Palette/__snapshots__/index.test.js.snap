// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette shoud match component Palette snapshot 1`] = `ShallowWrapper {}`;
